### PR TITLE
Set `CARGO` instead of `PATH` for Rust Clippy

### LINKED ISSUE #123227: set `CARGO` instead of `PATH`
Previously, clippy was using `cargo` from `PATH`, but since https://github.com/rust-lang/rust-clippy/pull/11944, it now prioritizes checking `CARGO` first. We can now set `cargo` path into `CARGO` and avoid appending it to the `PATH` chain.

In the bootstrap, we have this piece of code https://github.com/rust-lang/rust/blob/7f1d08e575a6578c9a591f9a78ab13deb33704f8/src/bootstrap/src/core/builder.rs#L1220-L1226 which adds `cargo` into `PATH`. As indicated by the FIXME note, https://github.com/rust-lang/rust-clippy/pull/11944 wasn't merged at the time. Now that we can do so, set `CARGO` instead of `PATH` and remove the FIXME note.

### PR DESCRIPTION
**Pull request by Rajveer100**
Resolves #123227

Previously, clippy was using `cargo` from `PATH`, but since [PR](https://github.com/rust-lang/rust-clippy/pull/11944), it now prioritises checking `CARGO` first.

### CODE REVIEW

File: src/bootstrap/src/core/builder.rs [hunk 1]
```diff
@@ -1196,11 +1196,10 @@ impl<'a> Builder<'a> {
     pub fn cargo_clippy_cmd(&self, run_compiler: Compiler) -> Command {
         let initial_sysroot_bin = self.initial_rustc.parent().unwrap();
         // Set PATH to include the sysroot bin dir so clippy can find cargo.
-        // FIXME: once rust-clippy#11944 lands on beta, set `CARGO` directly instead.
         let path = t!(env::join_paths(
             // The sysroot comes first in PATH to avoid using rustup's cargo.
             std::iter::once(PathBuf::from(initial_sysroot_bin))
-                .chain(env::split_paths(&t!(env::var("PATH"))))
+                .chain(env::split_paths(&t!(env::var("CARGO"))))
```
Comment (onur-ozkan) [inline src/bootstrap/src/core/builder.rs:1202]: You can set `CARGO` directly on the `cmd` right below the `let mut cmd` line. Sample ref:https://github.com/rust-lang/rust/blob/76cf07d5df52c07c3cd4cfeea1ab32b1cfba71bf/src/bootstrap/src/core/builder.rs#L1251

**Comment by onur-ozkan** [comment]
Thank you for working on this, but the current changes still don't seem correct. Here is the patch for the correct approach:

```diff
From c18be9da0560954cef2bde9e0fffa6ed215c9435 Mon Sep 17 00:00:00 2001
From: onur-ozkan <work@onurozkan.dev>
Date: Wed, 3 Apr 2024 14:37:47 +0300
Subject: [PATCH] remove `PATH` and set `CARGO` for `clippy` cmd

Signed-off-by: onur-ozkan <work@onurozkan.dev>
---
 src/bootstrap/src/core/builder.rs | 13 ++-----------
 1 file changed, 2 insertions(+), 11 deletions(-)

diff --git a/src/bootstrap/src/core/builder.rs b/src/bootstrap/src/core/builder.rs
index c051b818328..0cce5130f64 100644
--- a/src/bootstrap/src/core/builder.rs
+++ b/src/bootstrap/src/core/builder.rs
@@ -1194,20 +1194,11 @@ pub fn rustdoc(&self, compiler: Compiler) -> PathBuf {
     }
 
     pub fn cargo_clippy_cmd(&self, run_compiler: Compiler) -> Command {
-        let initial_sysroot_bin = self.initial_rustc.parent().unwrap();
-        // Set PATH to include the sysroot bin dir so clippy can find cargo.
-        // FIXME: once rust-clippy#11944 lands on beta, set `CARGO` directly instead.
-        let path = t!(env::join_paths(
-            // The sysroot comes first in PATH to avoid using rustup's cargo.
-            std::iter::once(PathBuf::from(initial_sysroot_bin))
-                .chain(env::split_paths(&t!(env::var("PATH"))))
-        ));
-
         if run_compiler.stage == 0 {
             // `ensure(Clippy { stage: 0 })` *builds* clippy with stage0, it doesn't use the beta clippy.
             let cargo_clippy = self.build.config.download_clippy();
             let mut cmd = Command::new(cargo_clippy);
-            cmd.env("PATH", &path);
+            cmd.env("CARGO", &self.initial_cargo);
             return cmd;
         }
 
@@ -1227,7 +1218,7 @@ pub fn cargo_clippy_cmd(&self, run_compiler: Compiler) -> Command {
 
         let mut cmd = Command::new(cargo_clippy);
         cmd.env(helpers::dylib_path_var(), env::join_paths(&dylib_path).unwrap());
-        cmd.env("PATH", path);
+        cmd.env("CARGO", &self.initial_cargo);
         cmd
     }
 
-- 
2.44.0
```

**Comment by Rajveer100** [comment]
Thanks @onur-ozkan!

**Pull request merged**